readme.md
# JSONP Serializer

[![Build status](https://img.shields.io/travis/phapi/serializer-jsonp.svg?style=flat-square)](https://travis-ci.org/phapi/serializer-jsonp)
[![Code Climate](https://img.shields.io/codeclimate/github/phapi/serializer-jsonp.svg?style=flat-square)](https://codeclimate.com/github/phapi/serializer-jsonp)
[![Test Coverage](https://img.shields.io/codeclimate/coverage/github/phapi/serializer-jsonp.svg?style=flat-square)](https://codeclimate.com/github/phapi/serializer-jsonp/coverage)

The JSONP Serializer package contains only one middleware, for serialization. The JSONP serializer works perfectly with the [JSON deserializer](https://github.com/phapi/serializer-json).

The serializer reacts if the <code>Accept</code> header matches one of the supported mime types.

By default the supported mime types are: <code>application/javascript</code> and <code>text/javascript</code>. It is possible to add more mime types by passing an array to the constructor.

## Installation
This middleware is **not** included by default in the [Phapi Framework](https://github.com/phapi/phapi-framework) but if you need to install it it's available to install via [Packagist](https://packagist.org) and [Composer](https://getcomposer.org).

```shell
$ php composer.phar require phapi/serializer-jsonp:1.*
```

## Configuration
The serializer has two configuration options:
- The name of the request header that contains the callback function name. *The default header name is set to **X-Callback***:
- Additional mime/types that the serializer should support.

```php
<?php
use Phapi\Middleware\Serializer\Jsonp\Jsonp;

$pipeline->pipe(new Jsonp($callbackHeader = 'X-Callback', ['text/html']));
```

Note that the array with additional mime types passed to the constructor will be merged with the default settings.

See the [configuration documentation](http://phapi.github.io/docs/started/configuration/) for more information about how to configure the integration with the Phapi Framework.

## Callback header
Please note that if the defined callback header cant be found or isn't a valid function name the serializer will serialize the response as JSON without including the provided callback function name.

## HTTP Status code
The HTTP status code is included in the body if an error occurs and the HTTP status is changed to 200 since many/all clients will have problems handling a response with a HTTP status that isn't 200.

You can read more about the problem with JSONP and HTTP status codes [here](http://www.theguardian.com/info/developer-blog/2012/jul/16/http-status-codes-jsonp).

## Phapi
This middleware is a Phapi package used by the [Phapi Framework](https://github.com/phapi/phapi-framework). The middleware are also [PSR-7](https://github.com/php-fig/http-message) compliant and implements the [Phapi Middleware Contract](https://github.com/phapi/contract).

## License
Serializer JSONP is licensed under the MIT License - see the [license.md](https://github.com/phapi/serializer-jsonp/blob/master/license.md) file for details

## Contribute
Contribution, bug fixes etc are [always welcome](https://github.com/phapi/serializer-jsonp/issues/new).


changelog.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [1.0.1] - 2015-09-23
### Fixed
- Fix typos in documentation.
- Fixed bug where the serializer modified the response body even if the content type didn't match the serializers mime types

## [1.0.0] - 2015-09-01
- Initial release!


tests/Phapi/Middleware/Serializer/Jsonp/JsonpTest.php
<?php

namespace Phapi\Tests\Middleware\Serializer\Jsonp;

use Phapi\Middleware\Serializer\Jsonp\Jsonp;
use PHPUnit_Framework_TestCase as TestCase;

/**
 * @coversDefaultClass \Phapi\Middleware\Serializer\Jsonp
 */
class JsonpTest extends TestCase {

    public function testException()
    {
        $serializer = new Jsonp();
        $this->setExpectedException('\Phapi\Exception\InternalServerError', 'Could not serialize content to JSONP');
        $serializer->serialize([ 'key' => "\xB1\x31"]);
    }

    public function testSerializeNoCallback()
    {
        $serializer = new Jsonp();
        $this->assertEquals('{"key":"value","another key":"second value"}',
            $serializer->serialize(['key' => 'value', 'another key' => 'second value'])
        );
    }

    public function testInvokeWithoutCallbackAndNoError()
    {
        $serializer = new Jsonp();

        $request = \Mockery::mock('Psr\Http\Message\ServerRequestInterface');
        $request->shouldReceive('hasHeader')->with('X-Callback')->andReturn(false);

        $response = \Mockery::mock('Psr\Http\Message\ResponseInterface');
        $response->shouldReceive('getStatusCode')->andReturn(200);
        $response->shouldReceive('hasHeader')->with('Content-Type')->andReturn(true);
        $response->shouldReceive('getHeaderLine')->with('Content-Type')->andReturn('application/javascript');
        $response->shouldReceive('getUnserializedBody')->andReturn(['key' => 'value']);
        $response->shouldReceive('withBody')->with(\Mockery::type('Psr\Http\Message\StreamInterface'))->andReturnSelf();

        $next = function ($request, $response, $next) {
            return $response;
        };

        $serializer($request, $response, $next);
    }

    public function testInvokeWithCallbackAndNoError()
    {
        $serializer = new Jsonp();

        $request = \Mockery::mock('Psr\Http\Message\ServerRequestInterface');
        $request->shouldReceive('hasHeader')->with('X-Callback')->andReturn(true);
        $request->shouldReceive('getHeaderLine')->with('X-Callback')->andReturn('someFunction');

        $response = \Mockery::mock('Psr\Http\Message\ResponseInterface');
        $response->shouldReceive('getStatusCode')->andReturn(200);
        $response->shouldReceive('hasHeader')->with('Content-Type')->andReturn(true);
        $response->shouldReceive('getHeaderLine')->with('Content-Type')->andReturn('application/javascript');
        $response->shouldReceive('getUnserializedBody')->andReturn(['key' => 'value']);
        $response->shouldReceive('withBody')->with(\Mockery::type('Psr\Http\Message\StreamInterface'))->andReturnSelf();

        $next = function ($request, $response, $next) {
            return $response;
        };

        $serializer($request, $response, $next);
    }

    public function testInvokeWithCallbackAndNoErrorAndSerialize()
    {
        $serializer = new Jsonp();

        $request = \Mockery::mock('Psr\Http\Message\ServerRequestInterface');
        $request->shouldReceive('hasHeader')->with('X-Callback')->andReturn(true);
        $request->shouldReceive('getHeaderLine')->with('X-Callback')->andReturn('someFunction');

        $response = \Mockery::mock('Psr\Http\Message\ResponseInterface');
        $response->shouldReceive('getStatusCode')->andReturn(200);
        $response->shouldReceive('hasHeader')->with('Content-Type')->andReturn(true);
        $response->shouldReceive('getHeaderLine')->with('Content-Type')->andReturn('application/javascript');
        $response->shouldReceive('getUnserializedBody')->andReturn(['key' => 'value']);
        $response->shouldReceive('withBody')->with(\Mockery::type('Psr\Http\Message\StreamInterface'))->andReturnSelf();

        $next = function ($request, $response, $next) {
            return $response;
        };

        $serializer($request, $response, $next);

        $expected = 'someFunction({"key":"value","next":"another value"})';
        $array = ['key' => 'value', 'next' => 'another value'];

        $this->assertEquals($expected, $serializer->serialize($array));
    }

    public function dataProviderValidation()
    {
        return [
            [ 'for' , false ],
            [ 'hello' , true ],
            [ 'callback/Function' , false ],
        ];
    }

    /**
     * @dataProvider dataProviderValidation
     */
    public function testValidation($callback, $expected)
    {
        $serializer = new Jsonp();
        $this->assertEquals($expected, $serializer->isValidateCallback($callback));
    }

    public function testInvokeWithCallbackAndErrorAndSerialize()
    {
        $serializer = new Jsonp();

        $request = \Mockery::mock('Psr\Http\Message\ServerRequestInterface');
        $request->shouldReceive('hasHeader')->with('X-Callback')->andReturn(true);
        $request->shouldReceive('getHeaderLine')->with('X-Callback')->andReturn('someFunction');

        $response = \Mockery::mock('Psr\Http\Message\ResponseInterface');
        $response->shouldReceive('getStatusCode')->andReturn(500);
        $response->shouldReceive('hasHeader')->with('Content-Type')->andReturn(true);
        $response->shouldReceive('getHeaderLine')->with('Content-Type')->andReturn('application/javascript');
        $response->shouldReceive('getUnserializedBody')->andReturn(['error' => 'Internal Server Error']);
        $response->shouldReceive('withBody')->with(\Mockery::type('Psr\Http\Message\StreamInterface'))->andReturnSelf();

        $response->shouldReceive('withStatus')->with(200)->andReturnSelf();
        $response->shouldReceive('withUnserializedBody')->with([ 'error' => 'Internal Server Error', 'HttpStatus' => 500])->andReturnSelf();

        $next = function ($request, $response, $next) {
            return $response;
        };

        $serializer($request, $response, $next);
    }

    public function testNotCompatibleResponseObject()
    {
        $serializer = new Jsonp();

        $request = \Mockery::mock('Psr\Http\Message\ServerRequestInterface');
        $request->shouldReceive('hasHeader')->with('X-Callback')->andReturn(true);
        $request->shouldReceive('getHeaderLine')->with('X-Callback')->andReturn('someFunction');

        $response = \Mockery::mock('Psr\Http\Message\ResponseInterface');
        $response->shouldReceive('getStatusCode')->andReturn(500);
        $response->shouldReceive('hasHeader')->with('Content-Type')->andReturn(true);
        $response->shouldReceive('getHeaderLine')->with('Content-Type')->andReturn('application/javascript');
        //$response->shouldReceive('getUnserializedBody')->andReturn(['error' => 'Internal Server Error']);
        $response->shouldReceive('withBody')->with(\Mockery::type('Psr\Http\Message\StreamInterface'))->andReturnSelf();

        $response->shouldReceive('withStatus')->with(200)->andReturnSelf();
        $response->shouldReceive('withUnserializedBody')->with([ 'error' => 'Internal Server Error', 'HttpStatus' => 500])->andReturnSelf();

        $next = function ($request, $response, $next) {
            return $response;
        };

        $this->setExpectedException('\RuntimeException', 'Serializer could not retrieve unserialized body');
        $serializer($request, $response, $next);
    }
    
    public function testNoContentType()
    {
        // Serializer
        $serializer = new Jsonp();

        // Container
        $container = \Mockery::mock('Phapi\Contract\Di\Container');
        $container->shouldReceive('offsetGet')->with('acceptTypes')->andReturn([]);
        $container->shouldReceive('offsetSet')->with('acceptTypes', ['application/javascript', 'text/javascript']);
        $serializer->setContainer($container);

        $serializer->registerMimeTypes();

        // Request
        $request = \Mockery::mock('Psr\Http\Message\ServerRequestInterface');

        $response = \Mockery::mock('Psr\Http\Message\ResponseInterface');
        $response->shouldReceive('hasHeader')->with('Content-Type')->andReturn(true);
        $response->shouldReceive('getHeaderLine')->with('Content-Type')->andReturn('application/xml');
        //$response->shouldReceive('getUnserializedBody')->andReturn([ 'username' => 'phapi' ]);
        //$response->shouldReceive('withBody')->with(\Mockery::type('Psr\Http\Message\StreamInterface'))->andReturnSelf();

        $serializer($request, $response, function ($request, $response) {
            return $response;
        });
    }
}


src/Phapi/Middleware/Serializer/Jsonp/Jsonp.php
<?php

namespace Phapi\Middleware\Serializer\Jsonp;

use Phapi\Exception\InternalServerError;
use Phapi\Serializer\Serializer;
use Psr\Http\Message\ResponseInterface;
use Psr\Http\Message\ServerRequestInterface;
use Zend\Diactoros\Stream;

/**
 * Class Json
 *
 * Middleware that serializes the response body to JSONP
 *
 * @category Phapi
 * @package  Phapi\Middleware\Serializer\Jsonp
 * @author   Peter Ahinko <peter@ahinko.se>
 * @license  MIT (http://opensource.org/licenses/MIT)
 * @link     https://github.com/phapi/serializer-jsonp
 */
class Jsonp extends Serializer
{

    /**
     * Valid mime types
     *
     * @var array
     */
    protected $mimeTypes = [
        'application/javascript',
        'text/javascript'
    ];

    /**
     * Name of the callback header to be used
     *
     * @var null|string
     */
    private $callbackHeader = null;

    /**
     * The callback function name
     *
     * @var null|string
     */
    private $callback = null;

    /**
     * Regex used to validate the callback function name
     *
     * @var string
     */
    private $regex = '/^[a-zA-Z_$][0-9a-zA-Z_$]*(?:\[(?:"(?:\\\.|[^"\\\])*"|\'(?:\\\.|[^\'\\\])*\'|\d+)\])*?$/';

    /**
     * Reserved words used to validate the callback function name
     *
     * @var array
     */
    private $reserved = [
        'break', 'do', 'instanceof', 'typeof', 'case', 'else', 'new', 'var', 'catch', 'finally', 'return',
        'void', 'continue', 'for', 'switch', 'while', 'debugger', 'function', 'this', 'with', 'default',
        'if', 'throw', 'delete', 'in', 'try', 'class', 'enum', 'extends', 'super', 'const', 'export',
        'import', 'implements', 'let', 'private', 'public', 'yield', 'interface', 'package', 'protected',
        'static', 'null', 'true', 'false',
    ];

    /**
     * Override the parent serializer constructor to allow specifying
     * the name of the header containing the callback function name
     *
     * @param string $callbackHeader
     * @param null $mimeTypes
     */
    public function __construct($callbackHeader = 'X-Callback', $mimeTypes = null)
    {
        $this->callbackHeader = $callbackHeader;

        parent::__construct($mimeTypes);
    }

    /**
     * Override the parent serializer invoke method. The method takes care of the
     * needed functionality to get and validate the callback function name from the
     * specified header as well as handling the http status if an error has occurred.
     * The http status code is included in the body if an error occurs and the http
     * status is changed to 200 since many/all clients will have problems handling
     * a response with a http status that isn't 200.
     *
     * @link http://www.theguardian.com/info/developer-blog/2012/jul/16/http-status-codes-jsonp
     *
     * @param ServerRequestInterface $request
     * @param ResponseInterface $response
     * @param callable|null $next
     * @return ResponseInterface
     */
    public function __invoke(ServerRequestInterface $request, ResponseInterface $response, callable $next = null)
    {
        $response = $next($request, $response, $next);

        // Get response content type
        $contentType = $this->getContentType($response);

        // Check if the accept header matches this serializers mime types
        if (!in_array($contentType, $this->mimeTypes)) {
            // This serializer does not handle this mime type so there is nothing
            // left to do. Return response.
            return $response;
        }

        // Get callback function name from request header or query string
        if ($request->hasHeader($this->callbackHeader)) {
            $this->callback = $request->getHeaderLine($this->callbackHeader);
        }

        // Validate callback
        $this->callback = ($this->isValidateCallback($this->callback)) ? $this->callback: null;

        // Get status code
        $status = $response->getStatusCode();

        // Check if error
        if ($status !== 200) {

            // Check if the response has a method for getting the unserialized body since
            // it's not part of the default PSR-7 implementation.
            try {
                $unserializedBody = $response->getUnserializedBody();
            } catch (\Exception $e) {
                throw new \RuntimeException('Serializer could not retrieve unserialized body');
            }

            // Check if the body is an array and not empty
            if (is_array($unserializedBody) && !empty($unserializedBody)) {
                // Add HTTP status code to body if not already set
                if (!isset($unserializedBody['HttpStatus'])) {
                    $unserializedBody['HttpStatus'] = $status;
                }

                // Change response status code to 200
                $response = $response->withStatus(200);

                // Set the unserialized body to response
                $response = $response->withUnserializedBody($unserializedBody);

                // Try and encode the array to json
                $json = $this->serialize($unserializedBody);

                // Create a new body with the serialized content
                $body = new Stream('php://memory', 'w+');
                $body->write($json);

                // Add the body to the response
                $response = $response->withBody($body);
            }
        }

        return $response;
    }

    /**
     * Validate a callback function name using regex and a list of
     * reserved words.
     *
     * @param $callback
     * @return bool
     */
    public function isValidateCallback($callback)
    {
        foreach (explode('.', $callback) as $part) {
            if (in_array($part, $this->reserved)) {
                return false;
            }

            if (!preg_match($this->regex, $part)) {
                return false;
            }
        }

        return true;
    }

    /**
     * Serialize body to json
     *
     * @param array $unserializedBody
     * @return string
     * @throws InternalServerError
     */
    public function serialize(array $unserializedBody = [])
    {
        // Serialize body
        if (false === $json = json_encode($unserializedBody)) {
            // Encode failed, throw error
            throw new InternalServerError('Could not serialize content to JSONP');
        }

        // Return with callback if it was found, else just json string
        return ($this->callback !== null) ? $this->callback . '(' . $json . ')': $json;
    }
}
